pykg2tbl/service.py

```diff
@@ -36,19 +44,83 @@ def as_csv(self, fileoutputlocation: str, sep: str = ","):
             should be written to.
         :param sep: delimiter that should be used for writing the csv file.
         """
-        # open the file in the write mode
-        f = open(fileoutputlocation, "w", newline="")
-        # create the csv writer
-        writer = csv.DictWriter(f, self._data[0].keys(), delimiter=sep)
-        # write a row to the csv file
-        for row in self._data:
-            writer.writerow(row)
-        # close the file
-        f.close()
+        try:
+            data = self.to_dataframe()
+        except Exception as e:
+            log.error(e)
+            data = self.to_list()
+        if isinstance(data, pd.DataFrame):
+            data.to_csv(fileoutputlocation, sep=sep)
+        else:
+            # open the file in the write mode
+            with open(fileoutputlocation, "w", newline="") as f:
+                # create the csv writer
+                writer = csv.DictWriter(f, data[0].keys(), delimiter=sep)
+                # write a row to the csv file
+                for row in data:
+                    writer.writerow(row)
+
+    def to_list(self) -> List:
+        """
+        Returns the list of query responses
+
+        :return: List of query responses
+        :rtype: list
+        """
+        return self._data
+
+    def to_dict(self) -> dict:
```

Comment:
And where should I implement this?
QueryResult also needs these methods

Reply:
https://github.com/vliz-be-opsci/pykg2tbl/pull/12#discussion_r1202200006

I was not thinking of subclassing this at all

Reply:
ok, but then it does not make sense to make it subclass ABC either ?

I thought QueryResult was going to be the abstract interface akin to KGSource
even if we currently have one (dict-based) implementation I would not mind having that distinction

but it can also be introduced through refactoring I guess

just that I am not convinced all sources we might encounter are going to be benefiting converting their results to a dict first (in general)